refactor(ast): remove `#[non_exhaustive]` attr from `AstBuilder`

overlookmotel opened the pull request:
Partially revert #4925. That PR's description gave no explanation of why this attribute is desirable.

overlookmotel commented:
* **#5130** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5130?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

DonIsaac reviewed (commented):
This is needed if we want to record node/scope/symbol/reference counts during parsing. We could add a property here that stores those statistics without breaking consuming code. If we have no plans on adding fields to `AstBuilder`, then this is safe to remove.

overlookmotel commented:
We don't have plans currently to add fields to `AstBuilder` currently, so I think best to remove this for now. I also seem less concerned about breaking changes than you are!

overlookmotel committed:
refactor(ast): remove `#[non_exhaustive]` attr from `AstBuilder` (#5130)

Partially revert #4925. That PR's description gave no explanation of why this attribute is desirable.
